src/metabase/db/update_h2.clj

```diff
@@ -89,13 +89,14 @@
 
 (defn update-if-needed!
   "Updates H2 database at db-path from version 1.x to 2.x if jdbc-url points
-   to version 1 H2 database."
+  to version 1 H2 database."
   [jdbc-url]
-  (locking h2-lock
-    (when (= 1 (db-version jdbc-url))
-      (log/info "H2 v1 database detected, updating...")
-      (try
-        (update! jdbc-url)
-        (catch Exception e
-          (log/error "Failed to update H2 database:" e)
-          (throw e))))))
+  (when (h2-base-path jdbc-url)
+    (locking h2-lock
+      (when (= 1 (db-version! jdbc-url))
```

Comment:
Is it safe to check this before we take the lock?

An optimistic pattern might help here to prevent us needing to ever take the lock:

```clojure
(defn update-needed? [jdbc-url]
  (= 1 (db-version! jdbc-url)))

(when (upgrade-needed? jdbc-url)
  (locking h2-lock
    ;; the database may have been upgraded while we waited for the lock
    (when (upgrade-needed? jdbc-url)
      (try
        (update! jdbc-url)
          ...)
```

A small thing, but a good pattern to pick up for writing concurrent code.

Reply:
well I thought about this but didn't bake it in because I don't think we create data-source much, probably only once per instance, so not sure if it's really needed.